everestrs/everestrs_sys/everestrs_sys.cpp

```diff
@@ -0,0 +1,83 @@
+#include "everestrs_sys.hpp"
+
+#include <cstdlib>
+#include <stdexcept>
+
+#include "everestrs/lib.rs.h"
+
+namespace {
+
+std::unique_ptr<Everest::Everest> create_everest_instance(
+    const std::string& module_id, const Everest::RuntimeSettings& rs,
+    const Everest::Config& config) {
+  return std::make_unique<Everest::Everest>(
+      module_id, config, true /* FIXME */, rs.mqtt_broker_host,
+      rs.mqtt_broker_port, rs.mqtt_everest_prefix, rs.mqtt_external_prefix,
+      rs.telemetry_prefix, rs.telemetry_enabled);
+}
+
+std::unique_ptr<Everest::Config> create_config_instance(
+    const Everest::RuntimeSettings& rs) {
+  // FIXME (aw): where to initialize the logger?
+  Everest::Logging::init(rs.logging_config_file);
+  return std::make_unique<Everest::Config>(
+      rs.schemas_dir.string(), rs.config_file.string(), rs.modules_dir.string(),
+      rs.interfaces_dir.string(), rs.types_dir.string(), rs.mqtt_everest_prefix,
+      rs.mqtt_external_prefix);
+}
+
+JsonBlob json2blob(const json& j) {
+  // I did not find a way to not copy the data at least once here.
+  const std::string dumped = j.dump();
+  rust::Vec<uint8_t> vec;
+  vec.reserve(dumped.size());
+  std::copy(dumped.begin(), dumped.end(), std::back_inserter(vec));
+  return JsonBlob{vec};
+}
+
+}  // namespace
+
+Module::Module(const std::string& module_id, const std::string& prefix,
+               const std::string& config_file)
+    : module_id_(module_id),
+      rs_(prefix, config_file),
+      config_(create_config_instance(rs_)),
+      handle_(create_everest_instance(module_id, rs_, *config_)) {}
+
+JsonBlob Module::get_interface(rust::Str interface_name) const {
+  const auto& interface_def =
+      config_->get_interface_definition(std::string(interface_name));
+  return json2blob(interface_def);
+}
+
+JsonBlob Module::initialize() {
+  handle_->connect();
+  handle_->spawn_main_loop_thread();
+
+  const std::string& module_name =
+      config_->get_main_config().at(module_id_).at("module");
+  return json2blob(config_->get_manifests().at(module_name));
+}
+
+void Module::signal_ready(const char* obj,
```

Comment:
I'm wondering can't this already take a function pointer? so you pass here a lambda from rust code? This would make the code clearer since we can get rid of the obscure *char pointer. Otherwise it should be probably a void pointer to indicate that we're passing just some address

Reply:
1. As discussed in person, it is not possible to pass a pointer to a lambda to C, since the calling conventions of Rust are different than the ones of C. The only way forward is to pass a function pointer with C conventions that gets passed a pointer to a Rust object, which can be casted in Rust into the right type and then used. This is widely unsafe of course. 

2. unfortunately, cxx.rs only exposes `char*`, not any other pointer type. I added a comment. 🤷